### PR TITLE
NEW getComposer() for more efficient Packagist API usage

### PR DESCRIPTION
**Pull request by chillu**
From https://packagist.org/apidoc:

> This is the preferred way to access the data as it is always up to date, and dumped to static files so it is very efficient on our end.

I think it's important to include this method and nudge people towards using it,
as a courtesy towards Packagist who are providing this API for free :)

This particular API endpoint is a bit confusing, it is clearly designed to find a single package,
yet it can return other unrelated in the "packages" response key.
My theory is that it's used for internal performance reasons by composer ("sharding").
I've opted to leave this unaltered in the PHP API, rather than pick out the specific package.
This should make it easier to read Packagist's API docs and understand the PHP API of this project.

The new endpoint also overlaps with a different "packages" response signature in the popular.json endpoint,
so I've opted to detect the types based on available keys in the factory.

Note that there's hundreds of pre-existing phpcs failures which I didn't fix,
it's a bit hard to tell if I made it worse (hopefully not!)

### CODE REVIEW

File: src/Packagist/Api/Result/Factory.php [hunk 1]
```diff
@@ -17,15 +17,22 @@ class Factory
...
         if (isset($data['packages'])) {
-            return $this->createSearchResults($data['packages']);
+            $packageOrResult = current($data['packages']);
+            if (isset($packageOrResult['versions'])) {
+                // Used for /p/<package>.json
+                return $this->createPackagesResults($data['packages']);
+            } else {
+                // Used for /explore/popular.json
+                return $this->createSearchResults($data['packages']);
+            }
```
Comment (robbieaverill) [inline src/Packagist/Api/Result/Factory.php:35]:
```suggestion
            }
            // Used for /explore/popular.json
            return $this->createSearchResults($data['packages']);
```

**Review by robbieaverill — COMMENTED** [review]
The accessing of packages by index feels a little strange to me, do you think we could wrap that up a bit better in this SDK?

**Comment by chillu** [comment]
@robbieaverill Sorry I updated the description as well as the implementation a bit after your review (you're too fast!).
If you look at `Factory->createSearchResults()`, it also uses package names as array keys (without an SDK wrapper for the collection), so my addition feels consistent with that? The fact that you ask for a specific package, but get back a list of `packages` containing keys is a weirdness from the Packagist API which I'm carrying through to this API. Makes sense?

**Comment by robbieaverill** [comment]
Yep, sweet as. I've raised #64 to make this library's API a little stricter - my objective is remove things like arbitrary array access in favour of typed interfaces in future (even if it's a PackageCollection or something). This will be good for now though! Thanks for clarifying.

File: README.md [hunk 2]
```diff
@@ -73,6 +78,32 @@ printf(
 Package sylius/sylius. Modern ecommerce for Symfony2.
 ```
 
+#### Get composer details: {#get-composer-details}
+
+Similar to `get()`, but uses composer metadata which is Packagist's preferred
```
Comment (robbieaverill) [inline README.md:83]:
```suggestion
Similar to `get()`, but uses Composer metadata which is Packagist's preferred
```

File: src/Packagist/Api/Result/Factory.php [hunk 3]
```diff
@@ -17,15 +17,22 @@ class Factory
...
         if (isset($data['packages'])) {
-            return $this->createSearchResults($data['packages']);
+            $packageOrResult = $data['packages'][array_key_first($data['packages'])];
+            if (isset($packageOrResult['name'])) {
+                // Used for /explore/popular.json
+                return $this->createSearchResults($data['packages']);
+            } else {
+                // Used for /p/<package>.json
+                return $this->createComposerPackagesResults($data['packages']);
+            }
```
Comment (robbieaverill) [inline src/Packagist/Api/Result/Factory.php:35]:
```suggestion
            }
            // Used for /p/<package>.json
            return $this->createComposerPackagesResults($data['packages']);
```

**Review by robbieaverill — APPROVED** [review]
LGTM, thanks! Minor suggested edits which I will apply

**Comment by chillu** [comment]
Thanks for the review Robbie, I've applied the suggestions.

**Comment by chillu** [comment]
I'm just rebasing this PR on https://github.com/KnpLabs/packagist-api/pull/64 to make it easier for you. Fine with having that go into the 2.0 release instead.

**Comment by robbieaverill** [comment]
Thanks @chillu, I'll get that PR merged and rebase this one again shortly

**Comment by chillu** [comment]
OK rebased, has the same conflicts as the other PR now, but should give you a good base to merge.

**Comment by chillu** [comment]
@robbieaverill I've fixed an issue that might have been introduced through the additional strict typing in #64. Haven't checked the composer schema but I assume all `maintainer` properties would be optional. You might also want to check if you're implicitly requiring other values to be present in your PR which are actually optional in the composer schema.

**Comment by robbieaverill** [comment]
Thanks @chillu, you didn't to do that, but I appreciate it nonetheless! Yes, introducing strict types did raise a number of inconsistencies with the existing interface over the schema - I'm considering that as a good thing!

**Comment by robbieaverill** [comment]
Rebased again and merged by hand, thanks @chillu

**Comment by robbieaverill** [comment]
See f97886d

**Pull request closed without merging**